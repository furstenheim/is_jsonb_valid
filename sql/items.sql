-- a schema given for items
-- valid items
SELECT is_jsonb_valid('{"items":{"type":"integer"}}', '[1,2,3]');
-- wrong type of items
SELECT is_jsonb_valid('{"items":{"type":"integer"}}', '[1,"x"]');
-- ignores non-arrays
SELECT is_jsonb_valid('{"items":{"type":"integer"}}', '{"foo":"bar"}');
-- JavaScript pseudo-array is valid
SELECT is_jsonb_valid('{"items":{"type":"integer"}}', '{"0":"invalid","length":1}');
-- an array of schemas for items
-- correct types
SELECT is_jsonb_valid('{"items":[{"type":"integer"},{"type":"string"}]}', '[1,"foo"]');
-- wrong types
SELECT is_jsonb_valid('{"items":[{"type":"integer"},{"type":"string"}]}', '["foo",1]');
-- incomplete array of items
SELECT is_jsonb_valid('{"items":[{"type":"integer"},{"type":"string"}]}', '[1]');
-- array with additional items
SELECT is_jsonb_valid('{"items":[{"type":"integer"},{"type":"string"}]}', '[1,"foo",true]');
-- empty array
SELECT is_jsonb_valid('{"items":[{"type":"integer"},{"type":"string"}]}', '[]');
-- JavaScript pseudo-array is valid
SELECT is_jsonb_valid('{"items":[{"type":"integer"},{"type":"string"}]}', '{"0":"invalid","1":"valid","length":2}');
-- items and subitems
-- valid items
SELECT is_jsonb_valid('{"definitions":{"item":{"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/sub-item"},{"$ref":"#/definitions/sub-item"}]},"sub-item":{"type":"object","required":["foo"]}},"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"}]}', '[[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}]]');
-- too many items
SELECT is_jsonb_valid('{"definitions":{"item":{"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/sub-item"},{"$ref":"#/definitions/sub-item"}]},"sub-item":{"type":"object","required":["foo"]}},"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"}]}', '[[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}]]');
-- too many sub-items
SELECT is_jsonb_valid('{"definitions":{"item":{"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/sub-item"},{"$ref":"#/definitions/sub-item"}]},"sub-item":{"type":"object","required":["foo"]}},"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"}]}', '[[{"foo":null},{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}]]');
-- wrong item
SELECT is_jsonb_valid('{"definitions":{"item":{"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/sub-item"},{"$ref":"#/definitions/sub-item"}]},"sub-item":{"type":"object","required":["foo"]}},"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"}]}', '[{"foo":null},[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}]]');
-- wrong sub-item
SELECT is_jsonb_valid('{"definitions":{"item":{"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/sub-item"},{"$ref":"#/definitions/sub-item"}]},"sub-item":{"type":"object","required":["foo"]}},"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"}]}', '[[{},{"foo":null}],[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}]]');
-- fewer items is valid
SELECT is_jsonb_valid('{"definitions":{"item":{"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/sub-item"},{"$ref":"#/definitions/sub-item"}]},"sub-item":{"type":"object","required":["foo"]}},"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"}]}', '[[{"foo":null}],[{"foo":null}]]');
-- nested items
-- valid nested array
SELECT is_jsonb_valid('{"type":"array","items":{"type":"array","items":{"type":"array","items":{"type":"array","items":{"type":"number"}}}}}', '[[[[1]],[[2],[3]]],[[[4],[5],[6]]]]');
-- nested array with invalid type
SELECT is_jsonb_valid('{"type":"array","items":{"type":"array","items":{"type":"array","items":{"type":"array","items":{"type":"number"}}}}}', '[[[["1"]],[[2],[3]]],[[[4],[5],[6]]]]');
-- not deep enough
SELECT is_jsonb_valid('{"type":"array","items":{"type":"array","items":{"type":"array","items":{"type":"array","items":{"type":"number"}}}}}', '[[[1],[2],[3]],[[4],[5],[6]]]');
-- items with null instance elements
-- allows null elements
SELECT is_jsonb_valid('{"items":{"type":"null"}}', '[null]');
-- array-form items with null instance elements
-- allows null elements
SELECT is_jsonb_valid('{"items":[{"type":"null"}]}', '[null]');