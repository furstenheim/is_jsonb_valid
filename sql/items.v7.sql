-- a schema given for items
-- valid items
SELECT is_jsonb_valid_draft_v7('{"items":{"type":"integer"}}', '[1,2,3]');
-- wrong type of items
SELECT is_jsonb_valid_draft_v7('{"items":{"type":"integer"}}', '[1,"x"]');
-- ignores non-arrays
SELECT is_jsonb_valid_draft_v7('{"items":{"type":"integer"}}', '{"foo":"bar"}');
-- JavaScript pseudo-array is valid
SELECT is_jsonb_valid_draft_v7('{"items":{"type":"integer"}}', '{"0":"invalid","length":1}');
-- an array of schemas for items
-- correct types
SELECT is_jsonb_valid_draft_v7('{"items":[{"type":"integer"},{"type":"string"}]}', '[1,"foo"]');
-- wrong types
SELECT is_jsonb_valid_draft_v7('{"items":[{"type":"integer"},{"type":"string"}]}', '["foo",1]');
-- incomplete array of items
SELECT is_jsonb_valid_draft_v7('{"items":[{"type":"integer"},{"type":"string"}]}', '[1]');
-- array with additional items
SELECT is_jsonb_valid_draft_v7('{"items":[{"type":"integer"},{"type":"string"}]}', '[1,"foo",true]');
-- empty array
SELECT is_jsonb_valid_draft_v7('{"items":[{"type":"integer"},{"type":"string"}]}', '[]');
-- JavaScript pseudo-array is valid
SELECT is_jsonb_valid_draft_v7('{"items":[{"type":"integer"},{"type":"string"}]}', '{"0":"invalid","1":"valid","length":2}');
-- items with boolean schema (true)
-- any array is valid
SELECT is_jsonb_valid_draft_v7('{"items":true}', '[1,"foo",true]');
-- empty array is valid
SELECT is_jsonb_valid_draft_v7('{"items":true}', '[]');
-- items with boolean schema (false)
-- any non-empty array is invalid
SELECT is_jsonb_valid_draft_v7('{"items":false}', '[1,"foo",true]');
-- empty array is valid
SELECT is_jsonb_valid_draft_v7('{"items":false}', '[]');
-- items with boolean schemas
-- array with one item is valid
SELECT is_jsonb_valid_draft_v7('{"items":[true,false]}', '[1]');
-- array with two items is invalid
SELECT is_jsonb_valid_draft_v7('{"items":[true,false]}', '[1,"foo"]');
-- empty array is valid
SELECT is_jsonb_valid_draft_v7('{"items":[true,false]}', '[]');
-- items and subitems
-- valid items
SELECT is_jsonb_valid_draft_v7('{"definitions":{"item":{"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/sub-item"},{"$ref":"#/definitions/sub-item"}]},"sub-item":{"type":"object","required":["foo"]}},"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"}]}', '[[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}]]');
-- too many items
SELECT is_jsonb_valid_draft_v7('{"definitions":{"item":{"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/sub-item"},{"$ref":"#/definitions/sub-item"}]},"sub-item":{"type":"object","required":["foo"]}},"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"}]}', '[[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}]]');
-- too many sub-items
SELECT is_jsonb_valid_draft_v7('{"definitions":{"item":{"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/sub-item"},{"$ref":"#/definitions/sub-item"}]},"sub-item":{"type":"object","required":["foo"]}},"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"}]}', '[[{"foo":null},{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}]]');
-- wrong item
SELECT is_jsonb_valid_draft_v7('{"definitions":{"item":{"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/sub-item"},{"$ref":"#/definitions/sub-item"}]},"sub-item":{"type":"object","required":["foo"]}},"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"}]}', '[{"foo":null},[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}]]');
-- wrong sub-item
SELECT is_jsonb_valid_draft_v7('{"definitions":{"item":{"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/sub-item"},{"$ref":"#/definitions/sub-item"}]},"sub-item":{"type":"object","required":["foo"]}},"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"}]}', '[[{},{"foo":null}],[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}]]');
-- fewer items is valid
SELECT is_jsonb_valid_draft_v7('{"definitions":{"item":{"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/sub-item"},{"$ref":"#/definitions/sub-item"}]},"sub-item":{"type":"object","required":["foo"]}},"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"}]}', '[[{"foo":null}],[{"foo":null}]]');
-- nested items
-- valid nested array
SELECT is_jsonb_valid_draft_v7('{"type":"array","items":{"type":"array","items":{"type":"array","items":{"type":"array","items":{"type":"number"}}}}}', '[[[[1]],[[2],[3]]],[[[4],[5],[6]]]]');
-- nested array with invalid type
SELECT is_jsonb_valid_draft_v7('{"type":"array","items":{"type":"array","items":{"type":"array","items":{"type":"array","items":{"type":"number"}}}}}', '[[[["1"]],[[2],[3]]],[[[4],[5],[6]]]]');
-- not deep enough
SELECT is_jsonb_valid_draft_v7('{"type":"array","items":{"type":"array","items":{"type":"array","items":{"type":"array","items":{"type":"number"}}}}}', '[[[1],[2],[3]],[[4],[5],[6]]]');
-- single-form items with null instance elements
-- allows null elements
SELECT is_jsonb_valid_draft_v7('{"items":{"type":"null"}}', '[null]');
-- array-form items with null instance elements
-- allows null elements
SELECT is_jsonb_valid_draft_v7('{"items":[{"type":"null"}]}', '[null]');