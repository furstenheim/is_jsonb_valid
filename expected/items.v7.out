-- a schema given for items
-- valid items
SELECT is_jsonb_valid_draft_v7('{"items":{"type":"integer"}}', '[1,2,3]');
 is_jsonb_valid_draft_v7 
-------------------------
 t
(1 row)

-- wrong type of items
SELECT is_jsonb_valid_draft_v7('{"items":{"type":"integer"}}', '[1,"x"]');
 is_jsonb_valid_draft_v7 
-------------------------
 f
(1 row)

-- ignores non-arrays
SELECT is_jsonb_valid_draft_v7('{"items":{"type":"integer"}}', '{"foo":"bar"}');
 is_jsonb_valid_draft_v7 
-------------------------
 t
(1 row)

-- JavaScript pseudo-array is valid
SELECT is_jsonb_valid_draft_v7('{"items":{"type":"integer"}}', '{"0":"invalid","length":1}');
 is_jsonb_valid_draft_v7 
-------------------------
 t
(1 row)

-- an array of schemas for items
-- correct types
SELECT is_jsonb_valid_draft_v7('{"items":[{"type":"integer"},{"type":"string"}]}', '[1,"foo"]');
 is_jsonb_valid_draft_v7 
-------------------------
 t
(1 row)

-- wrong types
SELECT is_jsonb_valid_draft_v7('{"items":[{"type":"integer"},{"type":"string"}]}', '["foo",1]');
 is_jsonb_valid_draft_v7 
-------------------------
 f
(1 row)

-- incomplete array of items
SELECT is_jsonb_valid_draft_v7('{"items":[{"type":"integer"},{"type":"string"}]}', '[1]');
 is_jsonb_valid_draft_v7 
-------------------------
 t
(1 row)

-- array with additional items
SELECT is_jsonb_valid_draft_v7('{"items":[{"type":"integer"},{"type":"string"}]}', '[1,"foo",true]');
 is_jsonb_valid_draft_v7 
-------------------------
 t
(1 row)

-- empty array
SELECT is_jsonb_valid_draft_v7('{"items":[{"type":"integer"},{"type":"string"}]}', '[]');
 is_jsonb_valid_draft_v7 
-------------------------
 t
(1 row)

-- JavaScript pseudo-array is valid
SELECT is_jsonb_valid_draft_v7('{"items":[{"type":"integer"},{"type":"string"}]}', '{"0":"invalid","1":"valid","length":2}');
 is_jsonb_valid_draft_v7 
-------------------------
 t
(1 row)

-- items with boolean schema (true)
-- any array is valid
SELECT is_jsonb_valid_draft_v7('{"items":true}', '[1,"foo",true]');
 is_jsonb_valid_draft_v7 
-------------------------
 t
(1 row)

-- empty array is valid
SELECT is_jsonb_valid_draft_v7('{"items":true}', '[]');
 is_jsonb_valid_draft_v7 
-------------------------
 t
(1 row)

-- items with boolean schema (false)
-- any non-empty array is invalid
SELECT is_jsonb_valid_draft_v7('{"items":false}', '[1,"foo",true]');
 is_jsonb_valid_draft_v7 
-------------------------
 f
(1 row)

-- empty array is valid
SELECT is_jsonb_valid_draft_v7('{"items":false}', '[]');
 is_jsonb_valid_draft_v7 
-------------------------
 t
(1 row)

-- items with boolean schemas
-- array with one item is valid
SELECT is_jsonb_valid_draft_v7('{"items":[true,false]}', '[1]');
 is_jsonb_valid_draft_v7 
-------------------------
 t
(1 row)

-- array with two items is invalid
SELECT is_jsonb_valid_draft_v7('{"items":[true,false]}', '[1,"foo"]');
 is_jsonb_valid_draft_v7 
-------------------------
 f
(1 row)

-- empty array is valid
SELECT is_jsonb_valid_draft_v7('{"items":[true,false]}', '[]');
 is_jsonb_valid_draft_v7 
-------------------------
 t
(1 row)

-- items and subitems
-- valid items
SELECT is_jsonb_valid_draft_v7('{"definitions":{"item":{"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/sub-item"},{"$ref":"#/definitions/sub-item"}]},"sub-item":{"type":"object","required":["foo"]}},"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"}]}', '[[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}]]');
 is_jsonb_valid_draft_v7 
-------------------------
 t
(1 row)

-- too many items
SELECT is_jsonb_valid_draft_v7('{"definitions":{"item":{"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/sub-item"},{"$ref":"#/definitions/sub-item"}]},"sub-item":{"type":"object","required":["foo"]}},"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"}]}', '[[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}]]');
 is_jsonb_valid_draft_v7 
-------------------------
 f
(1 row)

-- too many sub-items
SELECT is_jsonb_valid_draft_v7('{"definitions":{"item":{"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/sub-item"},{"$ref":"#/definitions/sub-item"}]},"sub-item":{"type":"object","required":["foo"]}},"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"}]}', '[[{"foo":null},{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}]]');
 is_jsonb_valid_draft_v7 
-------------------------
 f
(1 row)

-- wrong item
SELECT is_jsonb_valid_draft_v7('{"definitions":{"item":{"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/sub-item"},{"$ref":"#/definitions/sub-item"}]},"sub-item":{"type":"object","required":["foo"]}},"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"}]}', '[{"foo":null},[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}]]');
 is_jsonb_valid_draft_v7 
-------------------------
 f
(1 row)

-- wrong sub-item
SELECT is_jsonb_valid_draft_v7('{"definitions":{"item":{"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/sub-item"},{"$ref":"#/definitions/sub-item"}]},"sub-item":{"type":"object","required":["foo"]}},"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"}]}', '[[{},{"foo":null}],[{"foo":null},{"foo":null}],[{"foo":null},{"foo":null}]]');
 is_jsonb_valid_draft_v7 
-------------------------
 f
(1 row)

-- fewer items is valid
SELECT is_jsonb_valid_draft_v7('{"definitions":{"item":{"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/sub-item"},{"$ref":"#/definitions/sub-item"}]},"sub-item":{"type":"object","required":["foo"]}},"type":"array","additionalItems":false,"items":[{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"},{"$ref":"#/definitions/item"}]}', '[[{"foo":null}],[{"foo":null}]]');
 is_jsonb_valid_draft_v7 
-------------------------
 t
(1 row)

-- nested items
-- valid nested array
SELECT is_jsonb_valid_draft_v7('{"type":"array","items":{"type":"array","items":{"type":"array","items":{"type":"array","items":{"type":"number"}}}}}', '[[[[1]],[[2],[3]]],[[[4],[5],[6]]]]');
 is_jsonb_valid_draft_v7 
-------------------------
 t
(1 row)

-- nested array with invalid type
SELECT is_jsonb_valid_draft_v7('{"type":"array","items":{"type":"array","items":{"type":"array","items":{"type":"array","items":{"type":"number"}}}}}', '[[[["1"]],[[2],[3]]],[[[4],[5],[6]]]]');
 is_jsonb_valid_draft_v7 
-------------------------
 f
(1 row)

-- not deep enough
SELECT is_jsonb_valid_draft_v7('{"type":"array","items":{"type":"array","items":{"type":"array","items":{"type":"array","items":{"type":"number"}}}}}', '[[[1],[2],[3]],[[4],[5],[6]]]');
 is_jsonb_valid_draft_v7 
-------------------------
 f
(1 row)

-- single-form items with null instance elements
-- allows null elements
SELECT is_jsonb_valid_draft_v7('{"items":{"type":"null"}}', '[null]');
 is_jsonb_valid_draft_v7 
-------------------------
 t
(1 row)

-- array-form items with null instance elements
-- allows null elements
SELECT is_jsonb_valid_draft_v7('{"items":[{"type":"null"}]}', '[null]');
 is_jsonb_valid_draft_v7 
-------------------------
 t
(1 row)

